### ISSUE TITLE
NewPipe ports on other platforms

### ISSUE THREAD

**Issue by phly95** [TeamNewPipe/NewPipe#1051]
- [ y] I carefully read the [contribution guidelines](https://github.com/TeamNewPipe/NewPipe/blob/HEAD/.github/CONTRIBUTING.md) and agree to them.
- [ y] I checked if the issue/feature exists in the latest version.

This app is great, and I would love to have it on my Windows PC, especially with some  features that do not officially exist on the browser version, such as shuffling all videos on a single channel and playback without loading the video (as to not waste data).

**Comment by Hashik-Donthineni** [comment]
I have this idea my self, I guess it's on everyone's mind. @theScrabi IDK how you wrote the code for parsing and all, I believe it is not generic and can be applied to non-Android applications, why can't we go ahead and build one for the desktop using JavaFX or SwingX?

**Comment by speednroll** [comment]
I just thought it would not be possible but I miss an application like that, especially because the official website of YouTube is very slow for old computers (like the one I have). Now, the problem is if in case they manage to carry this application to the desktop, will require low resources?

**Comment by speednroll** [comment]
anyway I think we should first think about this when newpipe for android is already mature enough and with all the promising features that are yet to come (including subtitles, livestream, etc) to think of something even bigger, and enough bugs there is here to think about developing more bugs 😂

**Comment by theScrabi** [comment]
> I just thought it would not be possible

It is possible. In fact NewPipe is designed to work on multiple systems.
[This](https://github.com/TeamNewPipe/np-cli) is an example of a CLI app that graps the stream link out of a youtube link. It works on any device that has network and java.

What you would need to do is embedet the [NewPipeExtractor](https://github.com/TeamNewPipe/NewPipeExtractor) in your code.

**Comment by theScrabi** [comment]
>  think we should first think about this when newpipe for android is already mature enough

If someone is willing to develop it than develop it :D. Just because some people fix the bugs it does not mean all have to do it.

**Comment by speednroll** [comment]
> [This](https://github.com/TeamNewPipe/np-cli) is an example of a CLI app that graps the stream link out of a youtube link

I saw that example before, but I was thinking about literally the same application of newpipe into a computer, that may cost a bit more?

**Comment by Hashik-Donthineni** [comment]
@speednroll To answer your question "Now, the problem is if in case they manage to carry this application to the desktop, will require low resources?" 
The answer can't be straight and it's highly subjective. If you see a Slack desktop application which is built on Electron framework, if we had built it using JavaFX it will be significantly faster, but, JavaFX applications have long start time due to JVM, as it is said so.

**Comment by theScrabi** [comment]
Another option was of cource to write a NewPipe WebApplication. That works with a REST frontend for the NewPipeExtractor.

**Comment by Cormak** [comment]
@phly95 There's already an app for desktops. It's called Youtube-dl and it has a GUI frontend. I personally use it with a player called mpv to play videos directly on my screen but its mostly for downloading the video and/or audio. It's a lot more mature than Newpipe and thus works more reliably. 

https://rg3.github.io/youtube-dl/index.html

https://github.com/MrS0m30n3/youtube-dl-gui

**Comment by theScrabi** [comment]
I know about youtube-dl :D I took it code examples from it to write NewPipe. But as you said yt-dl is mostly for downloading. NewPipe is or should be a full blown client.

**Comment by phly95** [comment]
There is an app called SMplayer that does youtube, though newpipe has far surpassed it in terms of features, as SMtube has little to no integration with playlists and audio streams etc.

**Comment by theScrabi** [comment]
Meybe minitube can be tweaked by us

**Comment by PrestonN** [comment]
Hi there everyone!  If posts like this aren't allowed, please let me know and I'll gladly remove it.

I've been keeping an eye on the topic of a desktop YouTube client from various places and I've noticed the demand for something like this (especially since I've been wanting something like this myself).

I wanted to wait a little bit longer before I mentioned anything, but since the topic is fresh, I'd like to show a few things to everybody.  I'm currently working on a privacy minded YouTube client for desktop that aims to accomplish a similar goal to NewPipe.  I've decided to call it FreeTube and my plan is to Open Source it once the core functionality is finished.

The app is built using Electron and uses the YouTube API to acquire data.  It doesn't look like the NewPipe Extractor supports Electron but I would be interested in implementing it in the future if I'm wrong.

Here are a few pictures with some information as a proof of concept:
https://imgur.com/a/53Ahu

I hope that anyone interested in making a desktop client would like to help me out when I get this released.  The plan is to have it ready for initial release in a few weeks.

Thanks!

**Comment by speednroll** [comment]
in my opinion, Electron is slow, it is reserved only for media computers and high resources. in that case I prefer to use the official YouTube page instead, since both would require the same amount of resources

**Comment by PrestonN** [comment]
I can't argue with that too much.  I understand that a few people have issues running Electron on their machines due to the resources that it uses.  While I am trying to keep it as fast as possible, my main goal with this application is privacy.  If my application runs at about the same speed as YouTube, but delivers a more private browsing experience, then I feel like I was successful.  I hope to do better than that of course but I'll have to see what happens when I get there.

Regardless, I respect your opinion and I understand why you think that.

**Comment by theScrabi** [comment]
@PrestonN isn't electron just some reshaped chrome browser? So what you are working on is essentially a website that is running as a desktop app? Or am I wrong.

So point is @speednroll is when he says that electorn is slow. However I thought about having some sort of NewPipe website Maybe then what you are working on could be interesting. Honestly I like it already :D 

For the connection betwean FreeTube and the NewPipeExtractor we could use the REST Api for the NewPipeExtractor that I suggested erlyer, and which I'm willing to implement once multyservice turns out to work well :)

**Comment by PrestonN** [comment]
@theScrabi you are right in that Electron is a desktop wrapper based on Chromium.  I do believe that they remove a few things from base Chromium in order to help it run faster, but I'm not 100% certain.

The main reason I'm using Electron versus a website is that the audience that I'm targeting will most likely remove cookies from their browsers fairly often, if they save them at all.  This will make it difficult for them to save their subscriptions without creating some sort of account.  I'd assume that this is a big reason as to why similar websites such as HookTube do not include such a feature.

Electron allows me to save data such as subscriptions to the user's local storage without having to store cookies through the browser.  This helps keep it separate from their browsing history and prevents any trackers that may occur.

Using the REST API for the NewPipeExtractor sounds promising.  I'll definitely want to take a look at it once it's ready on your end.  My plan for an initial release is to keep going in my current direction but I'd love to take a look at it later in the future.

**Comment by Serkan-devel** [comment]
Maybe the Linux version could be built as an Appimage. @probonopd might know

**Comment by probonopd** [comment]
Is there a Linux version @Serkan-devel? Where can we get it?

**Comment by Serkan-devel** [comment]
@probonopd it should be packaged as one first

**Comment by Hashik-Donthineni** [comment]
But, if you think about it... what's the whole point of building a desktop version... we use NewPipe cause we can play it in background and pop-up player... No tracking. I don't think it'll pan out good, it's better if we keep making this app better just my thought guys.

**Comment by theScrabi** [comment]
@Hashik-Donthineni you may use it this way. I'm not, so I was happy to see a desktop/web version however we need more developer. So that the once that develop the android app can continue to make it better. I agree to U on that point :)

**Comment by probonopd** [comment]
> @probonopd it should be packaged as one first

I am a little bit confused since I was under the impression that this was a native Android (Java) app. I don't know how to package Android apps for Linux desktops (besides shipping a full Android emulator/VM, which I'd consider overkill).

**Comment by theScrabi** [comment]
yes, in fact I've been working on embedeting java into cpp applications. However I would recomend to use it as a seperate programm.

**Comment by Nutomic** [comment]
Here's a link to FreeTube because it wasn't linked so far. Gonna give it a try :)

https://github.com/FreeTubeApp/FreeTube

**Comment by theScrabi** [comment]
@PrestonN FreeTube is ultra awesome :heart_eyes:. I Would love to see a NewPipe version of it :D.

**Comment by PrestonN** [comment]
@theScrabi thanks for the kind words, it means a lot coming from you. :)

There's still a ton a work that needs to be done with it.  It's missing a few key features that people would expect something like FreeTube to have if it was a stable release.  Regardless, the features it does have seem to work well and I've been very pleased with the response it has had so far.

I would love to get the NewPipe Extractor working on FreeTube as I'm currently stuck with the YouTube API for now.  Let me know if there's anything I can do to help/test with it.

**Comment by Hashik-Donthineni** [comment]
FreeTube IMO feels little unfinished, don't get me wrong it's amazing and a lot of work is put into it, I mean couldn't they just copy YouTube's theme? Feels weird using it. The idea and the implementation are excellent, but the UI feels humm.. unfinished is the word.

**Comment by theScrabi** [comment]
If you guys remember how NewPipe looked in the beginning you would support freetube further. It even looked a bit like freetube now.

@PrestonN As I mentioned, my plan is to wrap the extractor with a REST/Json API, so it can easily be accessed from any device/os/lang. I guess that was perfect for freetube. However that rest wrapper just had to be written some time :)

**Comment by theScrabi** [comment]
So for the rest api this would be a good starting point:
https://jersey.github.io/

**Comment by Serkan-devel** [comment]
Isn't there a java API for qt?

**Comment by theScrabi** [comment]
The reason why i thin REST is good is because with that wed me more flexiple than only exposing the extractor to c++

**Comment by Hashik-Donthineni** [comment]
What about GraphQL? I did work with REST but not with GraphQL, heard it's good...

**Comment by theScrabi** [comment]
Well i through about god old ugly JSON, because it's good supported by many platforms, and simple to debug.

But if we want to discus about other protocols we can do that. Maybe we also want to consider [Protocoll buffers](https://developers.google.com/protocol-buffers/)

**Comment by theScrabi** [comment]
@TheAssassin @PrestonN what is your opinion about this?

**Comment by TheAssassin** [comment]
@theScrabi what do you need feedback for?

**Comment by theScrabi** [comment]
GraphQL

**Comment by TheAssassin** [comment]
GraphQL is overkill in my opinion.

**Comment by PrestonN** [comment]
@theScrabi Using REST would work better only because FreeTube is already structured around that and I'd be able to work on implementing the extractor quicker.  GraphQL should work though if that's the route you go with.

**Comment by TheAssassin** [comment]
A standard JSON RESTful API will probably suffice, yes. I doubt that the extra level of control (i.e., data filtering) etc. that GraphQL allows is needed. And querying is easier, too.

**Comment by theScrabi** [comment]
I don't know the way you query with GraphQL is also quite strait forward. I quite like it.
However what really matters is what someone implements, and if the implementation is good we'll take it.

**Comment by Hashik-Donthineni** [comment]
There is this Google Functions ( or Firebase Functions, I don't remember)   in Firebase it's still in beta I think we can use it for the API?

**Comment by theScrabi** [comment]
No firebase.

**Comment by theScrabi** [comment]
The reason is that NewPipe should be crept google independent. Therfore we should try to use as less google sevices as possible.

**Comment by Serkan-devel** [comment]
They are working on a universal web push api instead but it's just a draft by now https://developer.mozilla.org/en-US/docs/Web/API/Push_API

**Comment by Hashik-Donthineni** [comment]
Yes, I understand. I think it's about time someone starts on this, I think NewPipe is good enough on the mobile. I think it's hard to find people who work on the desktop version.

**Comment by theScrabi** [comment]
@Hashik-Donthineni thats why i am open about who implements it with what. @PrestonN already has something. That's a beginning.

I think once we have a working prototype people will come.

**Comment by theScrabi** [comment]
>  and use a native desktop player, like vlc?

That would be something I'd like to propose anyway.

**Comment by theScrabi** [comment]
I did. I like it.

**Comment by PrestonN** [comment]
> (Hint: We still need a REST or GraphQL frontend for the extractor ;D )

I'm guessing this means that there hasn't been too much work towards this? I'd be more than happy to use the extractor with FreeTube when it's ready.  The recent stuff with HookTube has definitely made me eager to get started and get rid of YouTube's API.  I have plenty of things to do in the mean time until it's ready though so no rush. :)

**Comment by theScrabi** [comment]
Well at the moment I'm still more concernet about the stability of the Android app, and the Multyservice project. I don't have the time and the resources to also make a desktop project at the moment. If we want to make this happen Someone had to take care about the interface.

**Comment by wb9688** [comment]
>(Hint: We still need a REST or GraphQL frontend for the extractor ;D )

Why would we need that for a desktop app? We could just directly use NewPipeExtractor (or through JNI from C/C++), right?

**Comment by theScrabi** [comment]
QT and java in the same executable can be done but I'd see pain in it.

Also I wanted to put a rest Api to it anyway, since it'll ad more flexibility, if we say ... We want to do a kody plugin, or a web version.

**Comment by wb9688** [comment]
@theScrabi: And from Python (or most other languages) you could call C/C++
code (which could call the Java code with JNI), there are libraries like
PyJNIus for that, but I don't know how good they work. Also not neccessary
Qt, GNOME (GTK+, GStreamer, etc.) is also an option, or even WxWidgets
(which uses GTK+/Win32/Cocoa depending on the OS) Or we could use a GUI
wrapper for Java, like java-gnome (which doesn't support Windows, so that
wouldn't be a good option). Btw a web version of NewPipe would be
impossible to do legally, because of CORS.

>

**Comment by TheAssassin** [comment]
For Python applications, Jython might help, as it has _direct_ access to the functions.

Any sort of "API bindings" is usually a PITA to maintain. I'm with @theScrabi here, it makes more sense to provide some sort of standalone API application that can be called via subprocesses, this is easier for everyone. All the data is then properly serialized (e.g., to JSON), and can be deserialized by consuming applications.

**Comment by wb9688** [comment]
@TheAssassin: Like D-Bus (does that work on non-Linux platforma though?)?I
totally agree, I was just listing other options. I do think calling a HTTP
API isn't the best option for the desktop app though.

Op za 28 jul. 2018 19:58 schreef TheAssassin <notifications@github.com>:

**Comment by TheAssassin** [comment]
No, just by creating a subprocess that provides an HTTP API. This is very simple, and works fine cross-platform.

OCS-Store, the client application for https://opendesktop.org stores, works like that. See https://www.opendesktop.org/p/1175480/.

Possible workflow:

  - main app calculates random free port (`bind()` to it to see whether it's free for use)
  - main app `fork()`s (and gets the forked process's ID)
  - forked process sets port in its process environment (e.g., `setenv("PORT", "12345")`)
  - forked process `exec()`s NewPipe Extractor API process
  - main UI now use HTTP API from `http://localhost:port`, and can monitor and manage subprocess

**Comment by wb9688** [comment]
@TheAssassin: So if I understand you correctly you want something like the following example?

![Diagram](https://user-images.githubusercontent.com/9780201/43359705-1a8d0138-92a7-11e8-9ae4-8135815d0535.png)

**Comment by TheAssassin** [comment]
Well, "want", it's just a proposal. But yeah, this is pretty much what I suggested, conceptually.

The magical part is the negotiation of the port that is used for the API. This is the most crucial part in such an architecture. Using fixed ports basically prevents using multiple instances.

**Comment by theScrabi** [comment]
> No, just by creating a subprocess that provides an HTTP API. This is very simple, and works fine cross-platform.

Excactly. :)

> And from Python (or most other languages) you could call C/C++ code (which could call the Java code with JNI)

Which means you have to use 3 Languages in one ... process or executable. Internal screaming.

**Comment by milkmiruku** [comment]
A related aside for Linux users only - [Anbox](https://anbox.io/), which allows one to run Android applications on any GNU/Linux operating system.

**Comment by 54D-R0807** [comment]
How about a complete port to Qt? You can compile to all major platforms.

**Comment by Serkan-devel** [comment]
I think that has been suggested earlier

**Comment by paulie-g** [comment]
The problem with a port is that someone will need to maintain it. Youtube changes things that break extraction. Writing a scraper for youtube is trivial in most languages with mature scraping infrastructure, the maintenance is the problem ergo the value is in reusing something that's already being maintained.

**Comment by TheAssassin** [comment]
@paulie-g the extractor itself was extracted to be able to further reduce the amount of dependencies. It's quite portable, AFAIK. I suppose the frontend might be a bit of an issue (I guess a desktop version is pretty much a separate code base), so I guess someone should take care of that as a project separate from NewPipe.

**Comment by theScrabi** [comment]
As the assasine said. The NewPipr extractor is written independent from Android. Actually for testing we already let it running on a PC. So if you want to write a Desktop app you only have to call the functions provided by the Extractor, and dont need to take care about Youtube at all.

**Comment by Hellzbellz123** [comment]
ill look into this subject on my own, but im not sure how far ill get

**Comment by paulie-g** [comment]
@TheAssassin @theScrabi You've either misunderstood me or misdirected your reply.

I am aware that Extractor has been made modular specifically to make it available for use in other projects. I think this is fantastic and shows that the devs cared enough to think ahead and put in the work so others could reuse their code in future - this is commendable. I think the NewPipe app itself is great, and the Extractor is the best full-scale youtube scraper (ytdl is limited to what is needed for downloads, everything else I've found works badly - especially yt api clients).

I, too, was interested in a desktop client, specifically a lightweight one - I often have dozens of videos open and paused at a time and don't want to pay for that in webengine memory usage. mps-youtube works well for music and I'd love to have a similar terminal client that has the features one needs for subscriptions, playlists, and so on. This is how I ended up reading this issue.

The point I was trying to make is that the value of Extractor is not just in the work that went into getting it to this point, but in the work the devs do in keeping it up to date. You can port it to [insert language here] at a given point in time, but even if you're able to make a very literal port - and that's not a foregone conclusion - you would still need to commit to continue porting all the updates in a timely manner to keep your client working. Therefore, unless that is a time commitment someone wants to make, it would be better to find a way to reuse the Extractor as is. That's all I was saying.

Incidentally - other than a compact, well-maintained port to something lightweight more in keeping with my own desires (c/rust/go) - I think packaging Extractor as-is in a server that can be fork()ed and talked to via a pipe is the best option. This way, whatever happens, any clients that use it will keep working as long as the NewPipe devs keep the original extractor up to date. Whether that can be made lightweight is another matter, and java is java, but that's not a requirement for most people (they may not want a 1+GB PSS of electron, but if QT is acceptable to them, so will running a jvm be).

**Comment by paulie-g** [comment]
Incidentally, the recommendation of using Avian (a lightweight jvm alternative w minimal class lib) seems reasonable for those who'd like to attempt this. I just submitted a PR for Avian that fixes compilation issues on modern Linux systems, so anyone who wants to play with it for this purpose can now do so (apply PR first - it won't be merged as Avian isn't actively maintained, but it works just fine).

**Comment by TheAssassin** [comment]
Recently, YouTube gets more and more annoying, even with lots of ads and JS blockers. I think there's more and more potential for a desktop version of NewPipe...

**Comment by Ashutosh-Kukreti** [comment]
I think freetube serves the purpose so there is no need for the desktop version of new pipe.
https://freetubeapp.io

**Comment by Serkan-devel** [comment]
but it relies on a 3rd-party service called invidious to function https://github.com/FreeTubeApp/FreeTube#how-does-it-work . If that service goes down but youtube doesn't, that client won't be usable

**Comment by Serkan-devel** [comment]
Also NewPipe is more than just an youtube client. One can also fetch content from soundcloud and from the ccc-library with it

**Comment by Ashutosh-Kukreti** [comment]
Freetube is still in beta ,if you want new features you can contact the developer and freetube relies on invidious to avoid the problem of updating the app (like in newpipe) every time youtube changes there api

**Comment by Ashutosh-Kukreti** [comment]
Also you can import new pipe subscriptions in freetube

**Comment by DI555** [comment]
I wanted to post the same issue also). And tryed to find some desktop app and didn'tfind it so comfortable like NP! Wait desktop version of NP so much! Is it built of java, yes? So imo won't be much difficulties.
Many thanks in advance!

**Comment by selurvedu** [comment]
@DI555, check out FreeTube or MiniTube.

**Comment by Isayso** [comment]
Just a thought. If I were Google, I wouldn't like an open source stand alone YT player which destroys my business plan. I would try to change things that bypass commercials would be impossible. They did this on web browsers already. 

I like the idea to have a newpipe on pc, but I'm afraid of waking up a sleeping dog, which would kill Newpipe on android as a side effect too.

**Comment by gkeegan** [comment]
@isayso I truly doubt YouTube is actively trying to do any of that, considering a simple adblocker can bypass ads for the millions of Firefox users who way outnumber NewPipe users. Anyways, FreeTube already exists and has not had any such issues.

**Comment by Isayso** [comment]
I hope you are right. Lets enjoy Newpipe. 
But you must add the Kodi users. I see the fight of the Video Kodi addons, they work often only a few weeks, Kodi itself fights against illegal streaming addons. Lawsuits from media companies against adblockers, new chrome development with less blocking capabilities.  We will see.

**Comment by gkeegan** [comment]
Kodi only has legal issues because the videos there are paid videos. NewPipe doesn't provide any paid videos. If it started giving access to YouTube Premium videos, then it would get shut down fairly quickly.

**Comment by Rik44444** [comment]
i would love a PC version that syncs with the Android app!

**Comment by Rik44444** [comment]
> 
> 
> Another option was of cource to write a NewPipe WebApplication. That works with a REST frontend for the NewPipeExtractor.

this is a great idea!! means a lot less extra development. + i would really love something that syncs my android playlists and favs to my laptop

**Comment by rkyoku** [comment]
Hi there! What about a PWA then? @theScrabi 
Or maybe people would rather a standalone app like a Qt or Go app (which would be ironic)?

**Comment by Stypox** [comment]
@rkyoku as a web app you can already use invidio.us

**Comment by rkyoku** [comment]
Hi there! What about a PWA then? @theScrabi  
Or maybe people would rather a fully standalone app like a Qt or Go app (which would be ironic)?
My personal opinion of (experienced) web dev would be to go the PWA route. Works litteraly everywhere with a single codebase. Client code remains visible, but I don't think this would be a problem.

**Comment by rkyoku** [comment]
@Stypox thanks for that! However, we are talking about expanding NewPipe here, and I thought I understood that theScrabi was hyped about that cross-platform app idea.

**Comment by Rik44444** [comment]
> 
> 
> @Stypox thanks for that! However, we are talking about expanding NewPipe here, and I thought I understood that theScrabi was hyped about that cross-platform app idea.

yes i agree, i use invidious too, but it would be great to have one of these video-watching applications (newpipe is my fav) be cross-platform. Also i think invidiou.us is only for Youtube, the great thing about NewPipe is that it is potentially for ANY video- or audio-sharing website. Atm it also includes Soundcloud and Peertube will also be included in the future!

**Comment by Stypox** [comment]
Yeah, I see what you mean
###### Btw, PeerTube was included very recently in 0.18.0 ;-)

**Comment by Rik44444** [comment]
> PWA

to add to this: i think a PWA is a really good idea. It appears to rapidly become the future for apps. + it takes less time to develop + cross-platform by design

**Comment by Rik44444** [comment]
> ###### Btw, PeerTube was included very recently in 0.18.0 ;-)
ah great! i hadn't seen that yet

**Comment by rkyoku** [comment]
> > PWA
> 
> to add to this: i think a PWA is a really good idea. It appears to rapidly become the future for apps. + it takes less time to develop + cross-platform by design

My point exactly :)

No promises here, but I *might* be able to allocate some time to lend a hand if this is the decided way to do it (and if help is needed @theScrabi ).

**Comment by 0alpha** [comment]
Existing solutions:
- https://github.com/ytdl-org/youtube-dl
- https://github.com/mpv-player/mpv (Optionally integrates with former)
- https://github.com/iina/iina (Uses mpv as a library, and thus can use ytdl)

Note: all above effectively require playlist/video url.

**Comment by sebma** [comment]
@theScrabi 
> > I just thought it would not be possible
> 
> It is possible. In fact NewPipe is designed to work on multiple systems.
> [This](https://github.com/TeamNewPipe/np-cli) is an example of a CLI app that graps the stream link out of a youtube link. It works on any device that has network and java.
> 
> What you would need to do is embedet the [NewPipeExtractor](https://github.com/TeamNewPipe/NewPipeExtractor) in your code.

Where is the **[np-cli](https://github.com/TeamNewPipe/np-cli)** now ? Has been removed ?

**Comment by grravity** [comment]
Big +1 for a Linux client

**Comment by yardenac** [comment]
> Where is the **[np-cli](https://github.com/TeamNewPipe/np-cli)** repository now ? Has been removed ?

There seems to be a [fork/mirror](https://github.com/goofwear/np-cli) but not sure how up-to-date it is. I'm also curious why it was removed.

**Comment by wb9688** [comment]
@yardenac: It's not up-to-date at all, plus np-cli was never really useful.

**Comment by petrhejna** [comment]
Hi, I like do big +1 to this and I happily contribute some Bitcoin to crowdfund NewPipe as Desktop App.

**Comment by GameyLP** [comment]
Uhm what about chatting with desktop client devs and making a standard output file and a option to sync over cloud (maybe nextcloud) that way people could sync newpipe with freetube and others could just join in and use the foundation for more applications like those?

**Comment by AndySchroder** [comment]
+1 for a linux client

**Comment by esotericDisciple** [comment]
++1 for a linux (desktop)  client, as all the available desktop app alternatives are either broken, missing essential features / downright ignoring user-friendliness, and/or are just electron wrappers that just show youtube.com...

**Comment by julian-fairfax** [comment]
Is there any alternative? I use this not just because of the ability to listen to watch youtube videos without ads but also because it's more private. I believe there are jailbreaks tweaks for the stock youtube app, but do you know if there's anything like this? It seems iOS lucks out in the open source software scene, not just because this project is kind of against some tos, but most open source software is just not supported on ios, which is a shame. Anyway, thank you so much for developing such a great app!

**Comment by Flam3z** [comment]
> Is there any alternative? I use this not just because of the ability to listen to watch youtube videos without ads but also because it's more private. I believe there are jailbreaks tweaks for the stock youtube app, but do you know if there's anything like this? It seems iOS lucks out in the open source software scene, not just because this project is kind of against some tos, but most open source software is just not supported on ios, which is a shame. Anyway, thank you so much for developing such a great app!

[FreeTube](https://freetubeapp.io/)

**Comment by julian-fairfax** [comment]
I don't see an iOS version of that.

**Comment by sebma** [comment]
@phly95 Can you please update your title to something like : "Windows/Linux/macOS Desktop Version?"

@julian-fairfax According to [this](https://github.com/FreeTubeApp/FreeTube/blob/development/README.md) : 

> FreeTube is an open source **desktop** YouTube player built with privacy in mind

Therefore it's not meant for mobile devices.

I suggest someone to create an separate issue for a NewPipe equivalent for other mobile OSes than Android (such as Windows Mobile/iOS/etc...)

**Comment by Atemu** [comment]
> Therefore it's not meant for mobile devices.

FreeTube's existence means that there is no need for a Desktop NP version anymore because that's essentially all a desktop NP could hope to be.  
I'd almost consider this issue fixed because of that.

I'd want syncing subscriptions, playlists, history etc. between FreeTube and NewPipe to be a thing before truly declaring it a desktop counterpart though.

> I suggest someone to create a separate issue for a NewPipe equivalent for other mobile OSes than Android (such as Windows Mobile/iOS/etc...)

The iOS App is still open of course but since the majority of NewPipe can't trivially be ported to iOS, I don't think it's worth opening an issue for that.  
An iOS App would be a completely separate project and the only part of NP that might be somewhat portable (NP Extractor) isn't even the most critical part as you could substitute it with youtube-dl for example.

Windows Mobile is a joke and also dead.

**Comment by martin-braun** [comment]
> I'd want syncing subscriptions, playlists, history etc. between FreeTube and NewPipe to be a thing before truly declaring it a desktop counterpart though.

@Atemu This, exactly this! It would be great if the devs of NewPipe could recognize the effort and willingness of @PrestonN go get this done, finally. The [issue](https://github.com/FreeTubeApp/FreeTube/issues/1026) exists on the FreeTube end and will be tackled by him when the devs of NewPipe willing to operate with him. 

We need a reliable privacy focused sync solution for all user data between NewPipe and FreeTube. This is the only alternative to a dedicated NewPipe desktop client.

**Comment by opusforlife2** [comment]
@martin-braun #3249 already exists for this.

And your comment seems to imply that the Freetube dev is stuck waiting on Newpipe team to do something, but that's not the case, is it? He himself has said that there are other issues with higher priority for him right now and he may _eventually_ get to this issue.

**Comment by martin-braun** [comment]
> @martin-braun #3249 already exists for this.
> 
> And your comment seems to imply that the Freetube dev is stuck waiting on Newpipe team to do something, but that's not the case, is it? He himself has said that there are other issues with higher priority for him right now and he may _eventually_ get to this issue.

@opusforlife2 Thank you for clarifying. Then I was missing some information and I'm sorry. If the NewPipe team opened all gates for a proper sync solution, everything is ok on the NewPipe's end. In this case, thank you to the devs of NewPipe for caring and taking action.

**Comment by opusforlife2** [comment]
@martin-braun I don't follow Freetube development. I just read through the issue you linked. Maybe you misinterpreted what the dev said?

Anyway, there has been no discussion between the Freetube and Newpipe devs for something like this yet.

**Comment by PrestonN** [comment]
I'll reiterate my stance from what I said in that thread.

I definitely want to look into the possibility of syncing between FreeTube and NewPipe at some point. Having said that, I'm at a point where I plan on focusing on several other things before I start any conversations. I'm not "waiting" on the NewPipe team for anything, I'm simply focusing on other priorities right now. This is all even assuming that the NewPipe team even _wants_ to consider syncing with FreeTube. I'm not saying that there's no interest in doing so but I couldn't force that decision even if I wanted it myself.

There's probably some smaller things we can discuss in the mean time, but it may be a while before I'd be ready to really start anything.

**Comment by Hei3h** [comment]
> * [y] Leí detenidamente las [pautas de contribución](https://github.com/TeamNewPipe/NewPipe/blob/HEAD/.github/CONTRIBUTING.md) y las acepto.
> * [y] Verifiqué si el problema / característica existe en la última versión.
> 
> Esta aplicación es excelente y me encantaría tenerla en mi PC con Windows, especialmente con algunas características que no existen oficialmente en la versión del navegador, como mezclar todos los videos en un solo canal y reproducirlos sin cargar el video (para no datos de residuos).

Could you please release a version in the Apple store, I understand that Google will never allow new Pipe in the play store, but I don't see why Apple wouldn't let new Pipe be in the Apple store

**Comment by selurvedu** [comment]
@Hei3h NewPipe was built for Android. It is tied to Android-specific libraries that don't exist outside of Android ecosystem. Hence it cannot be ported to other platforms without a complete rewrite (maybe except for its core part, NewPipeExtractor). If NewPipe was built with e.g. Flutter, Xamarin, React Native or some other cross-platform framework (I named a few popular ones that let you make apps for both Android and iOS from a single code base), then making a port would be feasible with reasonable effort.

**Comment by BeatLink** [comment]
Instead of a complete rewrite or building a desktop version, how about using a common filesystem format with FreeTube so that freetube can be the desktop counterpart? And data can be synced between the two apps using syncthing or any other filesystem based syncing software. Maybe both NewPipe and FreeTube can implement a [DecSync](https://github.com/39aldo39/DecSync)  backend so users can just sync between the two apps using syncthing

**Comment by BeatLink** [comment]
Instead of a complete rewrite or building a desktop version, how about using a common filesystem format with FreeTube so that freetube can be the desktop counterpart? And data can be synced between the two apps using syncthing or any other filesystem based syncing software. Maybe both NewPipe and FreeTube can implement a [DecSync](https://github.com/39aldo39/DecSync)  backend so users can just sync between the two apps using syncthing

**Comment by selurvedu** [comment]
@BeatLink this was discussed a few replies above.

**Comment by skyGtm** [comment]
It might be possible to run android apps in Windows 11
Related article here:
https://www.xda-developers.com/windows-11-android-apps/

**Comment by phly95** [comment]
Looks like Windows support's coming after all (albeit unofficially)

**Comment by Deagan** [comment]
For iOS users, check out Yattee, it looks to be a great alternative developing quickly!

https://github.com/yattee/yattee

**Comment by Bugginme** [comment]
For TVs follow this guide:
https://www.youtube.com/watch?app=desktop&v=_h9Cj_bl7rU
For desktops install ublock origin(gorhill) for browser

**Comment by eladkarako** [comment]
didn't saw anyone just pointing out to use yt-dlp,  
which is the most useful fork of youtube-dl,  
along with piping it to ffplay (or vlc for that matter).  
with a single thread downloading (normally it split download to 2-3 parts for faster download) you can start viewing the content pretty much right away, and it will work pretty well.  
there are pyinstall compiled builds for various os on their github releases page,  
that does not require you to even have python,  
and their integration of sponsorblock means you can skip the parts just like you do on a browser (just like the great https://github.com/polymorphicshade/NewPipe).  
and with the jsplayer signature decrypted properly, you won't suffer throttling.  
plus, unlike newpipe, their caching DOES NOT SUCK (yeah I've said it).

**Comment by Ahmadre** [comment]
Why don't we just use Flutter for all platforms 🤔?

**Comment by Ahmadre** [comment]
Why is my Flutter hint minimized? I'm honest to that suggestion. We develop desktop apps for linux, mac and windows and have professionel solutions made with Flutter.

**Comment by AudricV** [comment]
Sorry, your comment is not minimized anymore.

We do not want to use Flutter, as we do not have the bandwidth (too few developers very busy) to rewrite everything from scratch (the codebase has a strong integration with Android specific APIs), nor to support other platforms. It would also probably not bring anything better for the Android app: increase the Android app size a lot more, which is something we want to avoid (there is lightweight in the app description sentence 😅). See also discussions in #10118.

An unofficial port for Linux of the app has been released with our consent on Flathub, using Android Translation Layer: see #11597.

**Comment by Ahmadre** [comment]
> Sorry, your comment is not minimized anymore.
> 
> 
> 
> We do not want to use Flutter, as we do not have the bandwidth (too few developers very busy) to rewrite everything from scratch (the codebase has a strong integration with Android specific APIs), nor to support other platforms. It would also probably not bring anything better for the Android app: increase the Android app size a lot more, which is something we want to avoid (there is lightweight in the app description sentence 😅). See also discussions in #10118.
> 
> 
> 
> An unofficial port for Linux of the app has been released with our consent on Flathub, using Android Translation Layer: see #11597.

Awesome! Thanks for your links. I'll try to dig into this more :). Hava a wonderful weekend!

**Comment by paulie-g** [comment]
> Even if it's usable, the HW acceleration would'nt work, especially with dedicated GPUs like Nvidia. Freetube exists on linux, and it still doesn't support HW acceleration, making it unusable on high end desktops.

This is nonsensical. Just because one implementation can't do HW accel since it's non-native, it does not follow that all possible implementations will have the same problem. Simply using an external player like mpv gives you free HW accel, and I distinctly recall at least one QT-based youtube client that does this and many, many TUI-based ones that do.

**Comment by paulie-g** [comment]
> You can simply use yt-dlp as a backend to play everything in your favorite player that supports HW acceleration. 

I know what I can do, this is not a reddit how-to question. I don't care about the frontend.

> I would rather suggest Piped or Invidious devs to implement NewPipe extractor locally. Because YouTube is cracking down all the Piped/Invidious public servers as of now.

Porting the extractor is what I advocated for. Sharing it across all projects would mean paying the cost to maintain it with YT changes once. A lot of projects can benefit, but very few are prepared to lumber themselves with the JVM and the extractor isn't even a standalone lib. Hence the 'please port the extractor'.

**Comment by 12people** [comment]
Rather than start from scratch with Flutter, perhaps the app could slowly transition to [Jetpack Compose](https://developer.android.com/compose) for UI logic + [Kotlin Multiplatform](https://developer.android.com/kotlin/multiplatform) for business logic, and then ultimately to [Compose Multiplatform](https://www.jetbrains.com/compose-multiplatform/) -- all Android-native technologies?

That would:
- Allow a gradual, file-by-file transition, as old Java logic and XML UI code can live alongside the new code in one project.
- Allow a relatively straightforward transition, especially with business logic (where most Java code can be cleanly automatically translated to Kotlin).
- Make introducing new platforms much, much easier, as any other solution would require much larger rewrites.

**Comment by wizdom13** [comment]
For anyone still looking for a NewPipe-style desktop application: our independent NewPipe fork, [WizeStream](https://github.com/wizdom13/WizeStream), now includes a dedicated Desktop beta for Windows, macOS and Linux.

It supports browsing, playback, downloads, subscriptions, playlists, local media, Learning Mode notes and trusted-device synchronization. The current Desktop packages are unsigned beta builds.

WizeStream is an independent project and is not an official TeamNewPipe desktop port.
